modules/core/src/directx.cpp

```diff
@@ -1392,10 +1392,150 @@ void convertFromD3D11Texture2D(ID3D11Texture2D* pD3D11Texture2D, OutputArray dst
+                        clReleaseDevice(*device);
+                    }
+                    else
+                    {
+                        found = i;
+                    }
+                }
+            } while (0);
+
+            // try with CL_ALL_DEVICES_FOR_D3D11_KHR
+            if (found < 0) do {
+                device = NULL;
+                numDevices = 0;
+                status = clGetDeviceIDsFromD3D11KHR(platforms[i], CL_D3D11_DEVICE_KHR, pD3D11Device,
+                    CL_ALL_DEVICES_FOR_D3D11_KHR, 1, &(*device), &numDevices);
+                if (status != CL_SUCCESS)
+                    break;
+                if (numDevices > 0)
+                {
+                    cl_context_properties properties[] = {
+                            CL_CONTEXT_PLATFORM, (cl_context_properties)platforms[i],
+                            CL_CONTEXT_D3D11_DEVICE_KHR, (cl_context_properties)(pD3D11Device),
+                            CL_CONTEXT_INTEROP_USER_SYNC, CL_FALSE,
+                            NULL, NULL
+                    };
+                    context = clCreateContext(properties, 1, &(*device), NULL, NULL, &status);
+                    if (status != CL_SUCCESS)
+                    {
+                        clReleaseDevice(*device);
+                    }
+                    else
+                    {
+                        found = i;
+                    }
+                }
+            } while (0);
+
+            if (found >= 0) {
+                OpenCLExecutionContext clExecCtx;
+                try
+                {
+                    size_t out_size = 0;
+                    clGetDeviceInfo(*device, CL_DEVICE_EXTENSIONS, 0, nullptr, &out_size);
+                    std::vector<char> extensions(out_size);
+                    clGetDeviceInfo(*device, CL_DEVICE_EXTENSIONS, out_size, extensions.data(), nullptr);
+                    std::string ext(extensions.data(), out_size);
+                    if (!(ext.find("cl_khr_d3d11_sharing") != std::string::npos)) {
+                        CV_Error(cv::Error::OpenCLApiCallError, "OpenCL: Device doesn't contain cl_khr_d3d11_sharing extension");
+                    }
+                    clExecCtx = OpenCLExecutionContext::create(platformName, platform, context, *device);
+                    clExecCtx.getContext().setUserContext(std::make_shared<OpenCL_D3D11>(platform, pD3D11Device));
+                }
+                catch (...)
+                {
+                    clReleaseDevice(*device);
+                    clReleaseContext(context);
+                    throw;
+                }
+                clExecCtx.bind();
+                ctx = const_cast<Context&>(clExecCtx.getContext());
+                break;
+            }
+        }
     }
+
+    OpenCL_D3D11* impl = ctx.getUserContext<OpenCL_D3D11>().get();
+    cl_context* context = static_cast<cl_context*>(clContext);
+    *context = (cl_context)ctx.ptr();
+    if (nullptr == impl)
```

Comment:
`ctx` is a local variable - do we have guarantees that `*context` is not dangling in result ?